core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -242,10 +241,177 @@ public void messageReceived(final ReplicaRequest request, final TransportChannel
         }
     }
 
+    class PrimaryOperationTransportHandler implements TransportRequestHandler<Request> {
+
+        @Override
+        public void messageReceived(Request request, final TransportChannel channel) throws Exception {
+            ActionListener<Response> actionListener = new ActionListener<Response>() {
+                @Override
+                public void onResponse(Response response) {
+                    try {
+                        channel.sendResponse(response);
+                    } catch (Throwable t) {
+                        onFailure(t);
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    try {
+                        channel.sendResponse(e);
+                    } catch (Throwable t) {
+                        logger.warn("failed to send response for get", t);
+                    }
+                }
+            };
+            new AsyncPrimaryAction(request, actionListener).run();
+        }
+    }
+
+    class AsyncPrimaryAction extends AbstractRunnable {
+
+        private final ClusterStateObserver observer;
+        private final ActionListener<Response> listener;
+        private final InternalRequest internalRequest;
+        private final Request request;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
+        private volatile Releasable indexShardReference;
+
+        public AsyncPrimaryAction(Request request, ActionListener<Response> listener) {
+            this.observer = new ClusterStateObserver(clusterService, request.timeout(), logger);
+            this.listener = listener;
+            this.request = request;
+            this.internalRequest = new InternalRequest(request);
+            this.internalRequest.concreteIndex(request.internalShardRouting.getIndex());
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            finishAsFailed(t);
```

Comment:
Right, I see what you mean. This is a reason to put back the retry mechanism in `AsyncPrimaryAction`. Otherwise we would delay the write request for no reason, which isn't what we want.

But on the other hand then we have 2 retry mechanisms and we then also need a way of preventing that we retry twice. But some errors we really want only to retry on the coordination node, for example, if a the primary shard has just been removed on the node that did hold the primary and the primary shard is now somewhere else. In that case it is best to retry from the coordination node. So we need to properly deal with these kinds of situations.